### PR TITLE
JIT: Preserve range check for HW intrinsics with non-const/out-of-range immediates

### LINKED ISSUE #106546: JIT: Debug/Release diff for `AdvSimd.ShiftRightLogicalNarrowingLower`
Hit by Antigen:
```csharp
// Found by Antigen
// Reduced from 22.52 KB to 1.95 KB.


using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;
using System.Runtime.Intrinsics;
using System.Runtime.Intrinsics.Arm;
using System.Runtime.Intrinsics.X86;
using System.Numerics;
public class TestClass
{
    public struct S1
    {
    }
    public struct S2
    {
        public byte byte_1;
    }
    static Vector64<float> s_v64_float_25 = Vector64.Create(1f);
    static Vector128<long> s_v128_long_33 = Vector128<long>.AllBitsSet;
    static S1 s_s1_51 = new S1();
    static S2 s_s2_52 = new S2();
    byte byte_54 = 82;
    Vector64<byte> v64_byte_68 = Vector64.CreateScalar((byte)82);
    Vector64<int> v64_int_72 = Vector64.Create(1, 1);
    Vector64<double> v64_double_77 = Vector64.CreateScalar((double)82);
    S2 s2_103 = new S2();
    static int s_loopInvariant = 8;
    private static List<string> toPrint = new List<string>();
    public S2 Method3(S2 p_s2_147, Vector64<byte> p_v64_byte_148, S1 p_s1_149, ref S1 p_s1_150, double p_double_151, float p_float_152, ref byte p_byte_153)
    {
        unchecked
        {
            int __loopvar0 = s_loopInvariant;
            while (true)
            {
                if (__loopvar0 == s_loopInvariant + 9)
                    break;
                __loopvar0 += 3;
                Vector64.Shuffle(s_v64_float_25 += Vector64.AsSingle(v64_double_77), AdvSimd.ShiftRightLogicalNarrowingLower(s_v128_long_33, s2_103.byte_1) & (v64_int_72 - v64_int_72)* 15*4);
            }
            return s_s2_52;
        }
    }
    public void Method0()
    {
        unchecked
        {
            s_s2_52 = Method3(s2_103, v64_byte_68 -= v64_byte_68, s_s1_51, ref s_s1_51, 15-4, Vector64.Dot(Vector64<float>.AllBitsSet & s_v64_float_25, s_v64_float_25), ref byte_54);
            return;
        }
    }
    public static void Main(string[] args)
    {
        Console.WriteLine(Antigen());
    }
    public static int Antigen()
    {
        new TestClass().Method0();
        return string.Join(Environment.NewLine, toPrint).GetHashCode();
    }
}
/*
Environment:

export DOTNET_TieredPGO_InstrumentedTierAlwaysOptimized=1
export DOTNET_JitRandomOnStackReplacement=15
export DOTNET_HeapVerify=1
export DOTNET_TC_OnStackReplacement=1
export DOTNET_TC_PartialCompilation=1
export DOTNET_TC_QuickJitForLoops=1
export DOTNET_TieredCompilation=0
export DOTNET_TieredPGO=1
export DOTNET_EnableSSE2=0
export DOTNET_JitThrowOnAssertionFailure=1
export DOTNET_LegacyExceptionHandling=1

Release:
856641990

Debug:
Unhandled exception. System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
   at System.Runtime.Intrinsics.Arm.AdvSimd.ShiftRightLogicalNarrowingLower(Vector128`1 value, Byte count)
   at TestClass.Method3(S2 p_s2_147, Vector64`1 p_v64_byte_148, S1 p_s1_149, S1& p_s1_150, Double p_double_151, Single p_float_152, Byte& p_byte_153) in /Users/kunalpathak/git/repro/Program.cs:line 42
   at TestClass.Method0() in /Users/kunalpathak/git/repro/Program.cs:line 51
   at TestClass.Antigen() in /Users/kunalpathak/git/repro/Program.cs:line 61
   at TestClass.Main(String[] args) in /Users/kunalpathak/git/repro/Program.cs:line 57
*/
```

cc @dotnet/jit-contrib

### LINKED ISSUE #106608: JIT: `ArgumentOutOfRangeException` in debug but not release with `DotProductBySelectedQuadruplet`
```csharp
// Generated by Fuzzlyn v2.2 on 2024-08-17 15:13:27
// Run on Arm64 MacOS
// Seed: 17566447992392941035-vectort,vector64,vector128,armadvsimd,armadvsimdarm64,armaes,armarmbase,armarmbasearm64,armcrc32,armcrc32arm64,armdp,armrdm,armrdmarm64,armsha1,armsha256
// Reduced from 52.5 KiB to 0.7 KiB in 00:00:21
// Debug: Throws 'System.ArgumentOutOfRangeException'
// Release: Runs successfully
using System;
using System.Runtime.CompilerServices;
using System.Numerics;
using System.Runtime.Intrinsics;
using System.Runtime.Intrinsics.Arm;

public class Program
{
    public static void Main()
    {
        var vr7 = M1();
    }

    public static ushort M1()
    {
        var vr2 = Vector64.CreateScalar(0);
        var vr1 = AdvSimd.Extract(vr2, 0);
        var vr4 = Vector64.Create<int>(0);
        var vr5 = Vector64.Create<sbyte>(0);
        var vr6 = Vector64.Create<sbyte>(0);
        var vr3 = Dp.DotProductBySelectedQuadruplet(vr4, vr5, vr6, 7);
        return (ushort)((byte)ArmBase.Arm64.MultiplyHigh(vr1, 0) >> (-32768 & AdvSimd.Extract(vr3, 0)));
    }
}
```

cc @dotnet/jit-contrib

### PR DESCRIPTION
**Pull request by amanasifkhalid**
Fix #106608. Fix #106546. Based on discussion [here](https://github.com/dotnet/runtime/issues/106546#issuecomment-2300200301), intrinsics with out-of-range immediates should always be converted to user calls, and intrinsics with unknown immediates should be given range checks -- if we're optimizing and the immediate is later known, range check removal should kick in. For .NET 10, we plan to add more fallback intrinsic implementations so this pessimization doesn't kick in as often.

### CODE REVIEW

**Comment by amanasifkhalid** [comment]
cc @dotnet/jit-contrib, @tannergooding PTAL. Thanks!

**Comment by amanasifkhalid** [comment]
Looks like the `AdvSimd.StoreSelectedScalar` tests that try the maximum index are failing -- I suspect the JIT is calculating the wrong immediate bounds for this intrinsic. Taking a look...

**Comment by tannergooding** [comment]
The title `Always convert intrinsics with out-of-bounds immediates to user calls` is misleading/not quite correct.

What this is doing isn't converting to user calls, but rather just ensuring that a `GT_BOUNDS_CHECK` node is inserted, which ensures the required side effect exists in IR and persists. In the case the operand becomes an in range constant due to later constant folding, then the bounds check should get optimized away and the intrinsic will stay as an intrinsic, never getting rewritten to a user call.

**Comment by amanasifkhalid** [comment]
> What this is doing isn't converting to user calls, but rather just ensuring that a GT_BOUNDS_CHECK node is inserted, which ensures the required side effect exists in IR and persists.

The GT_BOUNDS_CHECK node is inserted when the immediate is unknown during importation, but if the immediate is known to be out-of-range (i.e. `CheckHWIntrinsicImmRange` returns false, and the immediate is constant) and we don't have a fallback, we bail in `Compiler::impHWIntrinsic`, which means it will become a user call, right?

**Comment by tannergooding** [comment]
> The GT_BOUNDS_CHECK node is inserted when the immediate is unknown during importation, but if the immediate is known to be out-of-range (i.e. CheckHWIntrinsicImmRange returns false, and the immediate is constant) and we don't have a fallback, we bail in Compiler::impHWIntrinsic, which means it will become a user call, right?

Ah, yes. I missed that bit of code, that will cause a user call to be emitted. We don't just import the throw directly (which would be more ideal) because the inliner wasn't happy with that.

The important part of the fix is that we're preserving the bounds check, which for most typical code will be due to values that aren't known to be constant during importation. The path where we have an out of bounds constant during import will likely never be hit in production (as users would have to knowingly pass in a bad value and ignore the analyzer warning).

**Review by tannergooding — APPROVED** [review]

File: src/coreclr/jit/hwintrinsic.cpp [hunk 1]
```diff
@@ -1332,29 +1332,26 @@ GenTree* Compiler::getArgForHWIntrinsic(var_types argType, CORINFO_CLASS_HANDLE
...
 {
     assert(immOp != nullptr);
     // Full-range imm-intrinsics do not need the range-check
     // because the imm-parameter of the intrinsic method is a byte.
     // AVX2 Gather intrinsics no not need the range-check
     // because their imm-parameter have discrete valid values that are handle by managed code
-    if (mustExpand && HWIntrinsicInfo::isImmOp(intrinsic, immOp)
+    if (!immOp->IsCnsIntOrI() && HWIntrinsicInfo::isImmOp(intrinsic, immOp)
 #ifdef TARGET_XARCH
         && !HWIntrinsicInfo::isAVX2GatherIntrinsic(intrinsic) && !HWIntrinsicInfo::HasFullRangeImm(intrinsic)
```
Comment (tannergooding) [inline src/coreclr/jit/hwintrinsic.cpp:1351]: We may want to explicitly check that `AVX2.Gather*` works correctly given its special.

The valid values are 1, 2, 4, and 8; other values should result in an exception.

Comment (amanasifkhalid) [reply src/coreclr/jit/hwintrinsic.cpp:1351]: Sure thing. I'll add a few more test cases.

**Comment by tannergooding** [comment]
> I'm nervous that similar intrinsics have similar logic issues, and our test coverage just isn't exposing them.

Looks like it is a bug that was introduced in .NET 9 as part of https://github.com/dotnet/runtime/pull/103689, https://github.com/dotnet/runtime/pull/93223, etc as part of the general work to expose the tuple overloads of the `Load/Store*` APIs. CC. @kunalspathak 

They were correctly encoded as `-1` in .NET 8 and we didn't have the potential issue where the second argument couldn't be used to discern the type due to being `TYP_STRUCT` then either.

**Comment by amanasifkhalid** [comment]
@tannergooding thanks for pointing that out. It's not the most elegant solution, but special-casing `StoreSelectedScalar` to always return the size of `Vector64` when the second argument is a `ValueTuple` seems to work. I don't know of any way to hard-code both 8 and 16 as its SIMD sizes in `hwintrinsiclistarm64.h` without creating another named intrinsic entry, so I think we want to keep looking up its SIMD size at runtime. If we look up the SIMD size using the first arg, `HWIntrinsicInfo::lookupSimdSize` will assert, because the first arg isn't an intrinsic type. If we use the second arg, the lookup works for all the `StoreSelectedScalar` invariants, except for the `ValueTuple` ones. I think going for a surgical fix at this point makes sense, unless there's some trivial way to handle the second arg being a non-intrinsic `TYP_STRUCT`.

**Comment by tannergooding** [comment]
> It's not the most elegant solution, but special-casing StoreSelectedScalar to always return the size of Vector64 when the second argument is a ValueTuple seems to work.

I don't think that's sufficient as there's overloads that take a tuple of `Vector128` in addition to the overloads that take a tuple of `Vector64`. It's also not sufficient o only cover `StoreSelectedScalar` as this would also impact `StoreVectorAndZip`, `Store`, `LoadAndInsertScalar`, and potentially some of the other `Load` variants that take a tuple.

> without creating another named intrinsic entry, so I think we want to keep looking up its SIMD size at runtime. If we look up the SIMD size using the first arg, HWIntrinsicInfo::lookupSimdSize will assert, because the first arg isn't an intrinsic type.

Right. The "proper" way is likely to have a flag on these intrinsics that lets us know we may need to check for `System.ValueTuple<...>`, get the type handle for the first generic type of the value tuple, and then pass that down to `getBaseJitTypeAndSizeOfSIMDType` instead.

You can see that some handling similar to this concept is being done for `LoadAndInsertScalar` already here: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/hwintrinsic.cpp#L1710-L1744 (gets the num instance fields, get the type of the field, then passes that into `getBaseJitTypeAndSizeOfSIMDType` to determine the `simdBaseType` and `simdSize`)

**Comment by amanasifkhalid** [comment]
@tannergooding thanks for pointing me in the right direction. I've added a new flag, `HW_Flag_BaseTypeFromValueTupleArg`, to denote this special case, and added handling for it to `impHWIntrinsic`. Do you know if any other intrinsics need this handling at the moment? I think it's just `StoreSelectedScalar`, since we can't hard-code its SIMD size. All the other ones you mentioned seem to be able to get away with that, or they have multiple named intrinsics for each variant (which can probably be cleaned up after this change goes through, though I don't know if it's worth changing the API surface).

File: src/coreclr/jit/hwintrinsiclistarm64.h [hunk 2]
```diff
@@ -487,7 +487,7 @@ HARDWARE_INTRINSIC(AdvSimd,       SignExtendWideningLower,
...
 HARDWARE_INTRINSIC(AdvSimd,       Store,                                                             8,      2,     {INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_invalid,        INS_invalid,        INS_st1_2regs,      INS_invalid},     HW_Category_MemoryStore,           HW_Flag_BaseTypeFromFirstArg|HW_Flag_SpecialImport|HW_Flag_SpecialCodeGen|HW_Flag_NeedsConsecutiveRegisters)
-HARDWARE_INTRINSIC(AdvSimd,       StoreSelectedScalar,                                               8,      3,     {INS_st1,            INS_st1,            INS_st1,            INS_st1,            INS_st1,            INS_st1,            INS_st1,            INS_st1,            INS_st1,            INS_st1},         HW_Category_MemoryStore,           HW_Flag_BaseTypeFromFirstArg|HW_Flag_HasImmediateOperand|HW_Flag_SIMDScalar|HW_Flag_SpecialCodeGen|HW_Flag_SpecialImport|HW_Flag_NeedsConsecutiveRegisters)
```
Comment (tannergooding) [inline src/coreclr/jit/hwintrinsiclistarm64.h:490]: There's several other methods that take `ValueTuple`, like `Store`, that need the same fix.

They aren't APIs taking an immediate, so I think it's fine to handle in a separate PR, but they do need to be handled.

**Review by tannergooding — APPROVED** [review]

File: src/coreclr/jit/hwintrinsiclistarm64.h (hunk 2 above)
Comment (amanasifkhalid) [reply src/coreclr/jit/hwintrinsiclistarm64.h:490]: I'll open a follow-up PR for those -- thanks

**Pull request merged**